### PR TITLE
REF/TYP: Rename ABCIndexClass->ABCIndex and use cast

### PR DESCRIPTION
**Pull request by rhshadrach**
- [ ] closes #xxxx
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/algorithms.py [hunk 1]
```diff
@@ -740,8 +739,11 @@ def factorize(
     uniques = _reconstruct_data(uniques, dtype, original)
 
     # return original tenor
-    if isinstance(original, ABCIndexClass):
+    if isinstance(original, ABCIndex):
         if original.dtype.kind in ["m", "M"] and isinstance(uniques, np.ndarray):
+            original._data = cast(
+                "Union[DatetimeArray, TimedeltaArray, np.ndarray]", original._data
+            )
```
Comment (jbrockmendel) [inline pandas/core/algorithms.py:746]: i think we can rule out ndarray here

Comment (rhshadrach) [reply pandas/core/algorithms.py:746]: Thanks, I took another look but I don't see why that's the case. Why is it `_data` can't be an ndarray?

Comment (jreback) [reply pandas/core/algorithms.py:746]: since this is a datetimelike subclass by-definition _data is always a datetimelike

**Comment by jbrockmendel** [comment]
@rhshadrach can you merge master

**Comment by jreback** [comment]
yep this looks good (just merge master as indicated) and ping on green.

**Comment by rhshadrach** [comment]
@jreback green.

**Comment by jreback** [comment]
thanks @rhshadrach 

IIRC there was an issue for this? or maybe just a comment

**Comment by rhshadrach** [comment]
To my knowledge this https://github.com/pandas-dev/pandas/pull/38055#issuecomment-733966956

**Pull request merged**